text/0000-inline-asm.md

```diff
@@ -0,0 +1,869 @@
+The only explicit operand is a register, that we fill from the variable `a`.
+The second implicit operand is the `eax` register which we fill from the variable `b`.
+The lower 32 bits of the result are stored in `eax` from which we fill the variable `lo`.
+The higher 32 bits are stored in `edx` from which we fill the variable `hi`.
+
+Note that `lateout` must be used for `eax` here since we are specifying the same register as both an input and an output.
+
+## Clobbered registers
+
+In many cases inline assembly will modify state that is not needed as an output.
+Usually this is either because we have to use a scratch register in the assembly,
+or instructions modify state that we don't need to further examine.
+This state is generally referred to as being "clobbered".
+We need to tell the compiler about this since it may need to save and restore this state
+around the inline assembly block.
+
+```rust
+let ebx: u32;
+let ecx: u32;
+
+unsafe {
+    asm!(
+        "cpuid",
+        in("eax") 4, in("ecx") 0,
+        lateout("ebx") ebx, lateout("ecx") ecx,
+        lateout("eax") _, lateout("edx") _
+    );
+}
+
+println!(
+    "L1 Cache: {}",
+    ((ebx >> 22) + 1) * (((ebx >> 12) & 0x3ff) + 1) * ((ebx & 0xfff) + 1) * (ecx + 1)
+);
+```
+
+In the example above we use the `cpuid` instruction to get the L1 cache size.
+This instruction writes to `eax`, `ebx`, `ecx`, and `edx`, but for the cache size we only care about the contents of `ebx` and `ecx`.
+
+However we still need to tell the compiler that `eax` and `edx` have been modified so that it can save any values that were in these registers before the asm. This is done by declaring these as outputs but with `_` instead of a variable name, which indicates that the output value is to be discarded.
+
+This can also be used with a general register class (e.g. `reg`) to obtain a scratch register for use inside the asm code.
+
+## Register template modifiers
+
+In some cases, fine control is needed over the way a register name is formatted when inserted into the template string. This is needed when an architecture's assembly language has several names for the same register, each typically being a "view" over a subset of the register (e.g. the low 32 bits of a 64-bit register).
+
+```rust
+let mut x: u16 = 0xab;
+
+unsafe {
+    asm!("mov {0:h} {0:b}", inout(reg_abcd) x);
+}
+
+assert_eq!(x, 0xabab);
+```
+
+In this example, we use the `reg_abcd` register class to restrict the register allocator to the 4 legacy x86 register (`ax`, `bx`, `cx`, `dx`) of which the first two bytes can be addressed independently.
+
+Let us assume that the register allocator has chosen to allocate `x` in the `ax` register.
+The `h` modifier will emit the register name for the high byte of that register and the `b` modifier will emit the register name for the low byte. The asm code will therefore be expanded as `mov ah, al` which copies the low byte of the value into the high byte.
```

Comment:
Shouldn't this be a `l` instead of `b` modifier?

Reply:
If you look at the [LLVM reference](https://llvm.org/docs/LangRef.html#asm-template-argument-modifiers), it says that the modifier is `b` for the low byte.